File: theory/src/main/kotlin/app/guitar/theory/CavaqSongs.kt
package app.guitar.theory

/**
 * Curated famous samba songs (CifraClub "mais acessadas / samba", top 30) tagged by the
 * FUNCTIONAL FAMILY of their core progression. Feeds the "Songs" button on the cavaquinho
 * Progressions screen: a taught [CavaqSequence] maps to one or more families, and
 * [forSequence] returns the songs in those families. Pure data — see
 * docs/cavaquinho-samba-songs.md for the per-song functional reductions (reviewable).
 */
data class CavaqSong(val title: String, val artist: String, val keyLabel: String, val family: String)

object CavaqSongs {

    val ALL: List<CavaqSong> = listOf(
        CavaqSong("O Mundo É Um Moinho", "Cartola", "A", "ii-V-I"),
        CavaqSong("Eu e Você Sempre", "Jorge Aragão", "E", "IV-iv"),
        CavaqSong("Meu Lugar", "Arlindo Cruz", "Am", "circle"),
        CavaqSong("Disritmia", "Martinho da Vila", "Gm", "circle"),
        CavaqSong("Nos Braços da Batucada", "Arlindo Cruz", "C", "I-vi-ii-V"),
        CavaqSong("Deixa a Vida Me Levar", "Zeca Pagodinho", "E", "ii-V-I"),
        CavaqSong("Trem das Onze", "Adoniran Barbosa", "Am", "minor-cadence"),
        CavaqSong("Fulminante", "Mumuzinho", "E", "I-IV-vamp"),
        CavaqSong("Lucidez", "Jorge Aragão", "E", "ii-V-I"),
        CavaqSong("Problema Emocional", "Reinaldo", "D", "circle"),
        CavaqSong("Preciso Me Encontrar", "Cartola", "Dm", "minor-cadence"),
        CavaqSong("Água de Chuva No Mar", "Beth Carvalho", "C", "ii-V-I"),
        CavaqSong("Retalhos de Cetim", "Benito Di Paula", "Am", "minor-cadence"),
        CavaqSong("Laços do Amor", "Grupo Fundo de Quintal", "C", "I-vi-ii-V"),
        CavaqSong("O Show Tem Que Continuar", "Grupo Fundo de Quintal", "C", "ii-V-I"),
        CavaqSong("Enredo Do Meu Samba", "Jorge Aragão", "A", "circle"),
        CavaqSong("Antigas Paixões", "Grupo Fundo de Quintal", "D", "IV-iv"),
        CavaqSong("Não Deixe O Samba Morrer", "Alcione", "Bm", "minor-cadence"),
        CavaqSong("Ezequiel 47", "Thiago Brito", "Gm", "minor-vamp"),
        CavaqSong("Você Me Vira a Cabeça", "Alcione", "F#m", "circle"),
        CavaqSong("Já é", "Jorge Aragão", "D", "ii-V-I"),
        CavaqSong("Se a Fila Andar", "Toninho Geraes", "Bb", "ii-V-I"),
        CavaqSong("Ah! Como Eu Amei", "Benito Di Paula", "Bb", "ii-V-I"),
        CavaqSong("O Bem", "Arlindo Cruz", "D", "I-vi-ii-V"),
        CavaqSong("Iracema", "Adoniran Barbosa", "Bm", "minor-cadence"),
        CavaqSong("Mais Feliz", "Zeca Pagodinho", "G", "IV-iv"),
        CavaqSong("As Rosas Não Falam", "Cartola", "Dm", "minor-cadence"),
        CavaqSong("Tiro ao Álvaro", "Adoniran Barbosa", "Bb", "I-vi-ii-V"),
        CavaqSong("Será Que É Amor", "Arlindo Cruz", "A", "ii-V-I"),
        CavaqSong("Carinhoso", "Pixinguinha", "G", "ii-V-I"),
    )

    /** Functional families each taught sequence (by id) covers. */
    private val SEQUENCE_FAMILIES: Map<String, Set<String>> = mapOf(
        "quadradinho_maj" to setOf("ii-V-I", "I-vi-ii-V"),
        "basic_min" to setOf("minor-cadence"),
        "medio_maj" to setOf("ii-V-I", "circle"),
        "medio_min" to setOf("minor-cadence", "circle"),
        "campo_maj" to emptySet(),
    )

    /** Curated songs whose functional family matches the given sequence id. */
    fun forSequence(sequenceId: String): List<CavaqSong> {
        val fams = SEQUENCE_FAMILIES[sequenceId] ?: emptySet()
        return ALL.filter { it.family in fams }
    }
}

File: chorect-web/src/theory/cavaqSongs.ts
// Curated famous samba songs (CifraClub "mais acessadas / samba", top 30) tagged by the
// FUNCTIONAL FAMILY of their core progression. Mirror of theory/CavaqSongs.kt. Feeds the
// "Songs" button on the cavaquinho Progressions screen. See docs/cavaquinho-samba-songs.md.

export interface CavaqSong { title: string; artist: string; keyLabel: string; family: string; }

export const CAVAQ_SONGS: CavaqSong[] = [
  { title: "O Mundo É Um Moinho", artist: "Cartola", keyLabel: "A", family: "ii-V-I" },
  { title: "Eu e Você Sempre", artist: "Jorge Aragão", keyLabel: "E", family: "IV-iv" },
  { title: "Meu Lugar", artist: "Arlindo Cruz", keyLabel: "Am", family: "circle" },
  { title: "Disritmia", artist: "Martinho da Vila", keyLabel: "Gm", family: "circle" },
  { title: "Nos Braços da Batucada", artist: "Arlindo Cruz", keyLabel: "C", family: "I-vi-ii-V" },
  { title: "Deixa a Vida Me Levar", artist: "Zeca Pagodinho", keyLabel: "E", family: "ii-V-I" },
  { title: "Trem das Onze", artist: "Adoniran Barbosa", keyLabel: "Am", family: "minor-cadence" },
  { title: "Fulminante", artist: "Mumuzinho", keyLabel: "E", family: "I-IV-vamp" },
  { title: "Lucidez", artist: "Jorge Aragão", keyLabel: "E", family: "ii-V-I" },
  { title: "Problema Emocional", artist: "Reinaldo", keyLabel: "D", family: "circle" },
  { title: "Preciso Me Encontrar", artist: "Cartola", keyLabel: "Dm", family: "minor-cadence" },
  { title: "Água de Chuva No Mar", artist: "Beth Carvalho", keyLabel: "C", family: "ii-V-I" },
  { title: "Retalhos de Cetim", artist: "Benito Di Paula", keyLabel: "Am", family: "minor-cadence" },
  { title: "Laços do Amor", artist: "Grupo Fundo de Quintal", keyLabel: "C", family: "I-vi-ii-V" },
  { title: "O Show Tem Que Continuar", artist: "Grupo Fundo de Quintal", keyLabel: "C", family: "ii-V-I" },
  { title: "Enredo Do Meu Samba", artist: "Jorge Aragão", keyLabel: "A", family: "circle" },
  { title: "Antigas Paixões", artist: "Grupo Fundo de Quintal", keyLabel: "D", family: "IV-iv" },
  { title: "Não Deixe O Samba Morrer", artist: "Alcione", keyLabel: "Bm", family: "minor-cadence" },
  { title: "Ezequiel 47", artist: "Thiago Brito", keyLabel: "Gm", family: "minor-vamp" },
  { title: "Você Me Vira a Cabeça", artist: "Alcione", keyLabel: "F#m", family: "circle" },
  { title: "Já é", artist: "Jorge Aragão", keyLabel: "D", family: "ii-V-I" },
  { title: "Se a Fila Andar", artist: "Toninho Geraes", keyLabel: "Bb", family: "ii-V-I" },
  { title: "Ah! Como Eu Amei", artist: "Benito Di Paula", keyLabel: "Bb", family: "ii-V-I" },
  { title: "O Bem", artist: "Arlindo Cruz", keyLabel: "D", family: "I-vi-ii-V" },
  { title: "Iracema", artist: "Adoniran Barbosa", keyLabel: "Bm", family: "minor-cadence" },
  { title: "Mais Feliz", artist: "Zeca Pagodinho", keyLabel: "G", family: "IV-iv" },
  { title: "As Rosas Não Falam", artist: "Cartola", keyLabel: "Dm", family: "minor-cadence" },
  { title: "Tiro ao Álvaro", artist: "Adoniran Barbosa", keyLabel: "Bb", family: "I-vi-ii-V" },
  { title: "Será Que É Amor", artist: "Arlindo Cruz", keyLabel: "A", family: "ii-V-I" },
  { title: "Carinhoso", artist: "Pixinguinha", keyLabel: "G", family: "ii-V-I" },
];

/** Functional families each taught sequence (by id) covers. */
const SEQUENCE_FAMILIES: Record<string, string[]> = {
  quadradinho_maj: ["ii-V-I", "I-vi-ii-V"],
  ii_v_i_maj: ["ii-V-I"],
  basic_min: ["minor-cadence"],
  medio_maj: ["ii-V-I", "circle"],
  campo_maj: [],
};

/** Curated songs whose functional family matches the given sequence id. */
export function cavaqSongsForSequence(sequenceId: string): CavaqSong[] {
  const fams = SEQUENCE_FAMILIES[sequenceId] ?? [];
  return CAVAQ_SONGS.filter((s) => fams.includes(s.family));
}
